Dockerfile

```diff
@@ -1,33 +1,10 @@
-FROM ubuntu:18.04
-LABEL maintainer="jdoliner@pachyderm.io"
-
-RUN \
-  apt-get update -yq && \
-  apt-get install -yq --no-install-recommends \
-    build-essential \
-    ca-certificates \
-    curl \
-    git \
-    libssl-dev \
-    pkg-config && \
-  apt-get clean && \
-  rm -rf /var/lib/apt
-COPY etc/compile/GO_VERSION GO_VERSION
-RUN \
-  curl -fsSL https://get.docker.com/builds/Linux/x86_64/docker-1.12.1.tgz | tar -C /bin -xz docker/docker --strip-components=1 && \
-  chmod +x /bin/docker
-RUN \
-  curl -sSL https://storage.googleapis.com/golang/$(cat GO_VERSION).linux-amd64.tar.gz | tar -C /tmp -xz && \
-  mkdir -p /usr/local/go && \
-  mv /tmp/go/bin /usr/local/go && \
-  mv /tmp/go/src /usr/local/go && \
-  mkdir -p /usr/local/go/pkg/tool/linux_amd64 && \
-  mv /tmp/go/pkg/include /usr/local/go/pkg && \
-  mv /tmp/go/pkg/tool/linux_amd64/compile /usr/local/go/pkg/tool/linux_amd64 && \
-  mv /tmp/go/pkg/tool/linux_amd64/asm /usr/local/go/pkg/tool/linux_amd64 && \
-  mv /tmp/go/pkg/tool/linux_amd64/link /usr/local/go/pkg/tool/linux_amd64 && \
-  rm -rf /tmp/go && \
-  mkdir -p /go/bin
-ENV PATH /go/bin:/usr/local/go/bin:$PATH
-ENV GOPATH /go
-RUN go get github.com/kisielk/errcheck golang.org/x/lint/golint
+# syntax=docker/dockerfile:1.0-experimental
+ARG GO_VERSION
+FROM golang:${GO_VERSION}
+WORKDIR /app
+COPY . .
+ARG LD_FLAGS
+RUN --mount=type=cache,target=/go/pkg/mod \
+    --mount=type=cache,target=/root/.cache/go-build \
+    CGO_ENABLED=0 go build -ldflags "${LD_FLAGS}" -o pachd "src/server/cmd/pachd/main.go" && \
+    CGO_ENABLED=0 go build -ldflags "${LD_FLAGS}" -o worker "src/server/cmd/worker/main.go"
```

Comment:
So it looks like this makes compilation of `worker` and `pachd` happen in serial rather than parallel like it was before. That might be part of the reason that you're not seeing things go faster with the caching changes. I think you could run them in parallel here.

I've also been wondering if maybe we should just merge the `pachd` and `worker` binaries, and have the `worker` mode just be another mode for `pachd`, we already have full mode and sidecar mode. I think that would simplify some things with our build and deploy process, and probably for users as well.

Reply:
The slowdown I had mentioned was specific to TravisCI. TravisCI has somewhat anemic support for caching docker images, so I have to naively save/load images to a cacheable directory. The old method cached all images, but because a full test run pulls so many images, it bloated the cache to 54gb, causing save/restore to slow down considerably more than the time saved. To get around this, I'm experimenting with selectively caching images.

I probably could parallelize go build commands, will look into that.

I tried merging the pachd/worker binaries earlier, but it doubles the size of the image of the pachd/worker images, since the `pachd` and `worker` binaries are the dominating factors wrt image size. Given that, I figured it was worthwhile to continue to keep them separate.

Reply:
That makes sense with regard to caching.

For combining the workers, did you actually combine them into a single binary, or put both binaries in the same image? I'd be kind of surprised if combining them leads to an image that's double in size, they should have a ton of code in common, and at the very least they share the same go runtime, which I think accounts for a meaningful amount of the size of the binary.

Reply:
Oh, I didn't consider combining them into a single binary -- I figured that'd be a too disruptive change.

Reply:
For what it's worth, even though the build process is serial, it's pretty fast because so much stuff is cached from the pachd build when the worker build runs. So build time shouldn't be a big consideration for combining the binaries -- though there are other benefits ofc.

Reply:
I think it's a little more disruptive, but probably not too bad. I'm actually more interested in the other benefits, I think it just simplifies things a bunch, means users wind up pulling less data, and probably makes it a little simpler for us to make sure we don't have a version mismatch between pachd and the worker. All that being said, I don't think this should be part of this PR, especially since the build times here isn't really an issue.